{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5d607ded",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b86275d5",
   "metadata": {},
   "source": [
    "# Genetic Variant Classification"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13de29f9",
   "metadata": {},
   "source": [
    "## Loading the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6667499a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_17271/4112643682.py:1: DtypeWarning: Columns (0,38,40) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv('data/clinvar_conflicting.csv')\n"
     ]
    }
   ],
   "source": [
    "data = pd.read_csv('data/clinvar_conflicting.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea8949ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>...</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1168180</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>0.1066</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.053</td>\n",
       "      <td>-0.208682</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1470752</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spinocerebellar_ataxia_21|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.000</td>\n",
       "      <td>6.517838</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1737942</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Human_Phenotype_Ontology:HP:0000486,MedGen:C00...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Strabismus|Nystagmus|Hypothyroidism|Intellectu...</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.100</td>\n",
       "      <td>6.061752</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.500</td>\n",
       "      <td>3.114491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.700</td>\n",
       "      <td>4.766224</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS REF ALT  AF_ESP  AF_EXAC  AF_TGP  \\\n",
       "0     1  1168180   G   C  0.0771  0.10020  0.1066   \n",
       "1     1  1470752   G   A  0.0000  0.00000  0.0000   \n",
       "2     1  1737942   A   G  0.0000  0.00001  0.0000   \n",
       "3     1  2160305   G   A  0.0000  0.00000  0.0000   \n",
       "4     1  2160305   G   T  0.0000  0.00000  0.0000   \n",
       "\n",
       "                                            CLNDISDB CLNDISDBINCL  \\\n",
       "0                                    MedGen:CN169374          NaN   \n",
       "1  MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...          NaN   \n",
       "2  Human_Phenotype_Ontology:HP:0000486,MedGen:C00...          NaN   \n",
       "3  MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...          NaN   \n",
       "4     MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002          NaN   \n",
       "\n",
       "                                               CLNDN  ...  \\\n",
       "0                                      not_specified  ...   \n",
       "1             Spinocerebellar_ataxia_21|not_provided  ...   \n",
       "2  Strabismus|Nystagmus|Hypothyroidism|Intellectu...  ...   \n",
       "3          Shprintzen-Goldberg_syndrome|not_provided  ...   \n",
       "4                       Shprintzen-Goldberg_syndrome  ...   \n",
       "\n",
       "                         SIFT           PolyPhen MOTIF_NAME MOTIF_POS  \\\n",
       "0                   tolerated             benign        NaN       NaN   \n",
       "1  deleterious_low_confidence             benign        NaN       NaN   \n",
       "2                 deleterious  probably_damaging        NaN       NaN   \n",
       "3                         NaN                NaN        NaN       NaN   \n",
       "4                         NaN                NaN        NaN       NaN   \n",
       "\n",
       "  HIGH_INF_POS MOTIF_SCORE_CHANGE  LoFtool  CADD_PHRED  CADD_RAW BLOSUM62  \n",
       "0          NaN                NaN      NaN       1.053 -0.208682      2.0  \n",
       "1          NaN                NaN      NaN      31.000  6.517838     -3.0  \n",
       "2          NaN                NaN      NaN      28.100  6.061752     -1.0  \n",
       "3          NaN                NaN      NaN      22.500  3.114491      NaN  \n",
       "4          NaN                NaN      NaN      24.700  4.766224     -3.0  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21e7bc3f",
   "metadata": {},
   "source": [
    "## Cleaning the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97b886ff",
   "metadata": {},
   "source": [
    "There are columns that can be dropped as they are not related to the problem at hand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "50771288",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.drop(columns = ['BAM_EDIT', 'EXON', 'CLNDISDB', 'CLNHGVS', \n",
    "         'MC', 'CLNVI', 'SYMBOL', 'Feature', 'Feature_type', \n",
    "         'BIOTYPE', 'INTRON', 'CADD_RAW'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f5662ff",
   "metadata": {},
   "source": [
    "#### Exploring nan values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6a9004f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>missing_percentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <td>0.999969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <td>0.999969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <td>0.999969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <td>0.999969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DISTANCE</th>\n",
       "      <td>0.998343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SSR</th>\n",
       "      <td>0.998006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLNSIGINCL</th>\n",
       "      <td>0.997438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <td>0.997438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLNDNINCL</th>\n",
       "      <td>0.997438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PolyPhen</th>\n",
       "      <td>0.619623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIFT</th>\n",
       "      <td>0.619010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BLOSUM62</th>\n",
       "      <td>0.607397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons</th>\n",
       "      <td>0.153464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Amino_acids</th>\n",
       "      <td>0.153464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein_position</th>\n",
       "      <td>0.152712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDS_position</th>\n",
       "      <td>0.152712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cDNA_position</th>\n",
       "      <td>0.136283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LoFtool</th>\n",
       "      <td>0.064628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <td>0.016752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRAND</th>\n",
       "      <td>0.000215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POS</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Consequence</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLASS</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORIGIN</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLNVC</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLNDN</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_TGP</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_EXAC</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_ESP</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALT</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMPACT</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    missing_percentage\n",
       "MOTIF_SCORE_CHANGE            0.999969\n",
       "HIGH_INF_POS                  0.999969\n",
       "MOTIF_POS                     0.999969\n",
       "MOTIF_NAME                    0.999969\n",
       "DISTANCE                      0.998343\n",
       "SSR                           0.998006\n",
       "CLNSIGINCL                    0.997438\n",
       "CLNDISDBINCL                  0.997438\n",
       "CLNDNINCL                     0.997438\n",
       "PolyPhen                      0.619623\n",
       "SIFT                          0.619010\n",
       "BLOSUM62                      0.607397\n",
       "Codons                        0.153464\n",
       "Amino_acids                   0.153464\n",
       "Protein_position              0.152712\n",
       "CDS_position                  0.152712\n",
       "cDNA_position                 0.136283\n",
       "LoFtool                       0.064628\n",
       "CADD_PHRED                    0.016752\n",
       "STRAND                        0.000215\n",
       "CHROM                         0.000000\n",
       "POS                           0.000000\n",
       "Consequence                   0.000000\n",
       "Allele                        0.000000\n",
       "CLASS                         0.000000\n",
       "ORIGIN                        0.000000\n",
       "CLNVC                         0.000000\n",
       "CLNDN                         0.000000\n",
       "AF_TGP                        0.000000\n",
       "AF_EXAC                       0.000000\n",
       "AF_ESP                        0.000000\n",
       "ALT                           0.000000\n",
       "REF                           0.000000\n",
       "IMPACT                        0.000000"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_perc = pd.DataFrame(data.isna().mean(), columns=['missing_percentage'])\n",
    "missing_perc=missing_perc.sort_values(by='missing_percentage', ascending=False)\n",
    "missing_perc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e81f906",
   "metadata": {},
   "source": [
    "We can drop the columns with over 99 percent missing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1659a23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['MOTIF_SCORE_CHANGE', 'HIGH_INF_POS', 'MOTIF_POS', 'MOTIF_NAME',\n",
       "       'DISTANCE', 'SSR', 'CLNSIGINCL', 'CLNDISDBINCL', 'CLNDNINCL'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_drop = missing_perc.index[:9]\n",
    "to_drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5a611031",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.drop(columns=to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "817318b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>CLNVC</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>...</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1168180</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>0.1066</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>522</td>\n",
       "      <td>174</td>\n",
       "      <td>E/D</td>\n",
       "      <td>gaG/gaC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.053</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1470752</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Spinocerebellar_ataxia_21|not_provided</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>509</td>\n",
       "      <td>170</td>\n",
       "      <td>P/L</td>\n",
       "      <td>cCg/cTg</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.000</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1737942</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Strabismus|Nystagmus|Hypothyroidism|Intellectu...</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>239</td>\n",
       "      <td>80</td>\n",
       "      <td>I/T</td>\n",
       "      <td>aTc/aCc</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.100</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome|not_provided</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>34</td>\n",
       "      <td>G/S</td>\n",
       "      <td>Ggc/Agc</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.500</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>34</td>\n",
       "      <td>G/C</td>\n",
       "      <td>Ggc/Tgc</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.700</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65183</th>\n",
       "      <td>X</td>\n",
       "      <td>154158201</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0801</td>\n",
       "      <td>0.13923</td>\n",
       "      <td>0.1605</td>\n",
       "      <td>Hereditary_factor_VIII_deficiency_disease|not_...</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3864</td>\n",
       "      <td>1288</td>\n",
       "      <td>S</td>\n",
       "      <td>tcA/tcC</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0.105</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65184</th>\n",
       "      <td>X</td>\n",
       "      <td>154159118</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>0.00060</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>not_specified|Hemophilia_A,_FVIII_Deficiency</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2947</td>\n",
       "      <td>983</td>\n",
       "      <td>V/I</td>\n",
       "      <td>Gta/Ata</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0.002</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65185</th>\n",
       "      <td>X</td>\n",
       "      <td>154194886</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0.00370</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>not_specified|Hemophilia_A,_FVIII_Deficiency</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1086</td>\n",
       "      <td>362</td>\n",
       "      <td>A</td>\n",
       "      <td>gcG/gcA</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>12.850</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65186</th>\n",
       "      <td>X</td>\n",
       "      <td>154490187</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.00034</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Non-syndromic_X-linked_intellectual_disability...</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>543</td>\n",
       "      <td>181</td>\n",
       "      <td>T</td>\n",
       "      <td>acA/acG</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.130</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65187</th>\n",
       "      <td>X</td>\n",
       "      <td>154508542</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0019</td>\n",
       "      <td>0.00267</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>not_specified|not_provided</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>532</td>\n",
       "      <td>178</td>\n",
       "      <td>P/A</td>\n",
       "      <td>Cca/Gca</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.14000</td>\n",
       "      <td>0.046</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65188 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS REF ALT  AF_ESP  AF_EXAC  AF_TGP  \\\n",
       "0         1    1168180   G   C  0.0771  0.10020  0.1066   \n",
       "1         1    1470752   G   A  0.0000  0.00000  0.0000   \n",
       "2         1    1737942   A   G  0.0000  0.00001  0.0000   \n",
       "3         1    2160305   G   A  0.0000  0.00000  0.0000   \n",
       "4         1    2160305   G   T  0.0000  0.00000  0.0000   \n",
       "...     ...        ...  ..  ..     ...      ...     ...   \n",
       "65183     X  154158201   T   G  0.0801  0.13923  0.1605   \n",
       "65184     X  154159118   C   T  0.0020  0.00060  0.0013   \n",
       "65185     X  154194886   C   T  0.0125  0.00370  0.0111   \n",
       "65186     X  154490187   T   C  0.0003  0.00034  0.0000   \n",
       "65187     X  154508542   G   C  0.0019  0.00267  0.0008   \n",
       "\n",
       "                                                   CLNDN  \\\n",
       "0                                          not_specified   \n",
       "1                 Spinocerebellar_ataxia_21|not_provided   \n",
       "2      Strabismus|Nystagmus|Hypothyroidism|Intellectu...   \n",
       "3              Shprintzen-Goldberg_syndrome|not_provided   \n",
       "4                           Shprintzen-Goldberg_syndrome   \n",
       "...                                                  ...   \n",
       "65183  Hereditary_factor_VIII_deficiency_disease|not_...   \n",
       "65184       not_specified|Hemophilia_A,_FVIII_Deficiency   \n",
       "65185       not_specified|Hemophilia_A,_FVIII_Deficiency   \n",
       "65186  Non-syndromic_X-linked_intellectual_disability...   \n",
       "65187                         not_specified|not_provided   \n",
       "\n",
       "                           CLNVC  ORIGIN  ...  CDS_position Protein_position  \\\n",
       "0      single_nucleotide_variant       1  ...           522              174   \n",
       "1      single_nucleotide_variant       1  ...           509              170   \n",
       "2      single_nucleotide_variant      35  ...           239               80   \n",
       "3      single_nucleotide_variant      33  ...           100               34   \n",
       "4      single_nucleotide_variant      33  ...           100               34   \n",
       "...                          ...     ...  ...           ...              ...   \n",
       "65183  single_nucleotide_variant       1  ...          3864             1288   \n",
       "65184  single_nucleotide_variant       1  ...          2947              983   \n",
       "65185  single_nucleotide_variant       1  ...          1086              362   \n",
       "65186  single_nucleotide_variant       1  ...           543              181   \n",
       "65187  single_nucleotide_variant       1  ...           532              178   \n",
       "\n",
       "      Amino_acids   Codons STRAND                        SIFT  \\\n",
       "0             E/D  gaG/gaC    1.0                   tolerated   \n",
       "1             P/L  cCg/cTg   -1.0  deleterious_low_confidence   \n",
       "2             I/T  aTc/aCc   -1.0                 deleterious   \n",
       "3             G/S  Ggc/Agc    1.0                         NaN   \n",
       "4             G/C  Ggc/Tgc    1.0                         NaN   \n",
       "...           ...      ...    ...                         ...   \n",
       "65183           S  tcA/tcC   -1.0                         NaN   \n",
       "65184         V/I  Gta/Ata   -1.0                   tolerated   \n",
       "65185           A  gcG/gcA   -1.0                         NaN   \n",
       "65186           T  acA/acG   -1.0                         NaN   \n",
       "65187         P/A  Cca/Gca   -1.0                         NaN   \n",
       "\n",
       "                PolyPhen  LoFtool CADD_PHRED  BLOSUM62  \n",
       "0                 benign      NaN      1.053       2.0  \n",
       "1                 benign      NaN     31.000      -3.0  \n",
       "2      probably_damaging      NaN     28.100      -1.0  \n",
       "3                    NaN      NaN     22.500       NaN  \n",
       "4                    NaN      NaN     24.700      -3.0  \n",
       "...                  ...      ...        ...       ...  \n",
       "65183                NaN  0.00158      0.105       NaN  \n",
       "65184             benign  0.00158      0.002       3.0  \n",
       "65185                NaN  0.00158     12.850       NaN  \n",
       "65186                NaN      NaN      0.130       NaN  \n",
       "65187                NaN  0.14000      0.046      -1.0  \n",
       "\n",
       "[65188 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "097e45af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 65188 entries, 0 to 65187\n",
      "Data columns (total 25 columns):\n",
      " #   Column            Non-Null Count  Dtype  \n",
      "---  ------            --------------  -----  \n",
      " 0   CHROM             65188 non-null  object \n",
      " 1   POS               65188 non-null  int64  \n",
      " 2   REF               65188 non-null  object \n",
      " 3   ALT               65188 non-null  object \n",
      " 4   AF_ESP            65188 non-null  float64\n",
      " 5   AF_EXAC           65188 non-null  float64\n",
      " 6   AF_TGP            65188 non-null  float64\n",
      " 7   CLNDN             65188 non-null  object \n",
      " 8   CLNVC             65188 non-null  object \n",
      " 9   ORIGIN            65188 non-null  int64  \n",
      " 10  CLASS             65188 non-null  int64  \n",
      " 11  Allele            65188 non-null  object \n",
      " 12  Consequence       65188 non-null  object \n",
      " 13  IMPACT            65188 non-null  object \n",
      " 14  cDNA_position     56304 non-null  object \n",
      " 15  CDS_position      55233 non-null  object \n",
      " 16  Protein_position  55233 non-null  object \n",
      " 17  Amino_acids       55184 non-null  object \n",
      " 18  Codons            55184 non-null  object \n",
      " 19  STRAND            65174 non-null  float64\n",
      " 20  SIFT              24836 non-null  object \n",
      " 21  PolyPhen          24796 non-null  object \n",
      " 22  LoFtool           60975 non-null  float64\n",
      " 23  CADD_PHRED        64096 non-null  float64\n",
      " 24  BLOSUM62          25593 non-null  float64\n",
      "dtypes: float64(7), int64(3), object(15)\n",
      "memory usage: 12.4+ MB\n"
     ]
    }
   ],
   "source": [
    "data.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4133e601",
   "metadata": {},
   "source": [
    "As the order of the rows is determined by first the chromosome that the variant is on and then the position of the allele, we will fill the missing values in the below rows with forward fill, as the best logical guess we can make is that the closer the position, the more related the values.\n",
    "\n",
    "For the other numerical values, we can interpolate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "84b4332c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>CLNVC</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>PolyPhen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>552</td>\n",
       "      <td>522</td>\n",
       "      <td>174</td>\n",
       "      <td>E/D</td>\n",
       "      <td>gaG/gaC</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Spinocerebellar_ataxia_21|not_provided</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>523</td>\n",
       "      <td>509</td>\n",
       "      <td>170</td>\n",
       "      <td>P/L</td>\n",
       "      <td>cCg/cTg</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Strabismus|Nystagmus|Hypothyroidism|Intellectu...</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>632</td>\n",
       "      <td>239</td>\n",
       "      <td>80</td>\n",
       "      <td>I/T</td>\n",
       "      <td>aTc/aCc</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>probably_damaging</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome|not_provided</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>132</td>\n",
       "      <td>100</td>\n",
       "      <td>34</td>\n",
       "      <td>G/S</td>\n",
       "      <td>Ggc/Agc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>132</td>\n",
       "      <td>100</td>\n",
       "      <td>34</td>\n",
       "      <td>G/C</td>\n",
       "      <td>Ggc/Tgc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65183</th>\n",
       "      <td>X</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Hereditary_factor_VIII_deficiency_disease|not_...</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>G</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>4035</td>\n",
       "      <td>3864</td>\n",
       "      <td>1288</td>\n",
       "      <td>S</td>\n",
       "      <td>tcA/tcC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65184</th>\n",
       "      <td>X</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>not_specified|Hemophilia_A,_FVIII_Deficiency</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>3118</td>\n",
       "      <td>2947</td>\n",
       "      <td>983</td>\n",
       "      <td>V/I</td>\n",
       "      <td>Gta/Ata</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65185</th>\n",
       "      <td>X</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>not_specified|Hemophilia_A,_FVIII_Deficiency</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>1257</td>\n",
       "      <td>1086</td>\n",
       "      <td>362</td>\n",
       "      <td>A</td>\n",
       "      <td>gcG/gcA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65186</th>\n",
       "      <td>X</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Non-syndromic_X-linked_intellectual_disability...</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>822</td>\n",
       "      <td>543</td>\n",
       "      <td>181</td>\n",
       "      <td>T</td>\n",
       "      <td>acA/acG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65187</th>\n",
       "      <td>X</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>not_specified|not_provided</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>791</td>\n",
       "      <td>532</td>\n",
       "      <td>178</td>\n",
       "      <td>P/A</td>\n",
       "      <td>Cca/Gca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65188 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM REF ALT                                              CLNDN  \\\n",
       "0         1   G   C                                      not_specified   \n",
       "1         1   G   A             Spinocerebellar_ataxia_21|not_provided   \n",
       "2         1   A   G  Strabismus|Nystagmus|Hypothyroidism|Intellectu...   \n",
       "3         1   G   A          Shprintzen-Goldberg_syndrome|not_provided   \n",
       "4         1   G   T                       Shprintzen-Goldberg_syndrome   \n",
       "...     ...  ..  ..                                                ...   \n",
       "65183     X   T   G  Hereditary_factor_VIII_deficiency_disease|not_...   \n",
       "65184     X   C   T       not_specified|Hemophilia_A,_FVIII_Deficiency   \n",
       "65185     X   C   T       not_specified|Hemophilia_A,_FVIII_Deficiency   \n",
       "65186     X   T   C  Non-syndromic_X-linked_intellectual_disability...   \n",
       "65187     X   G   C                         not_specified|not_provided   \n",
       "\n",
       "                           CLNVC Allele         Consequence    IMPACT  \\\n",
       "0      single_nucleotide_variant      C    missense_variant  MODERATE   \n",
       "1      single_nucleotide_variant      A    missense_variant  MODERATE   \n",
       "2      single_nucleotide_variant      G    missense_variant  MODERATE   \n",
       "3      single_nucleotide_variant      A    missense_variant  MODERATE   \n",
       "4      single_nucleotide_variant      T    missense_variant  MODERATE   \n",
       "...                          ...    ...                 ...       ...   \n",
       "65183  single_nucleotide_variant      G  synonymous_variant       LOW   \n",
       "65184  single_nucleotide_variant      T    missense_variant  MODERATE   \n",
       "65185  single_nucleotide_variant      T  synonymous_variant       LOW   \n",
       "65186  single_nucleotide_variant      C  synonymous_variant       LOW   \n",
       "65187  single_nucleotide_variant      C    missense_variant  MODERATE   \n",
       "\n",
       "      cDNA_position CDS_position Protein_position Amino_acids   Codons  \\\n",
       "0               552          522              174         E/D  gaG/gaC   \n",
       "1               523          509              170         P/L  cCg/cTg   \n",
       "2               632          239               80         I/T  aTc/aCc   \n",
       "3               132          100               34         G/S  Ggc/Agc   \n",
       "4               132          100               34         G/C  Ggc/Tgc   \n",
       "...             ...          ...              ...         ...      ...   \n",
       "65183          4035         3864             1288           S  tcA/tcC   \n",
       "65184          3118         2947              983         V/I  Gta/Ata   \n",
       "65185          1257         1086              362           A  gcG/gcA   \n",
       "65186           822          543              181           T  acA/acG   \n",
       "65187           791          532              178         P/A  Cca/Gca   \n",
       "\n",
       "                             SIFT           PolyPhen  \n",
       "0                       tolerated             benign  \n",
       "1      deleterious_low_confidence             benign  \n",
       "2                     deleterious  probably_damaging  \n",
       "3                             NaN                NaN  \n",
       "4                             NaN                NaN  \n",
       "...                           ...                ...  \n",
       "65183                         NaN                NaN  \n",
       "65184                   tolerated             benign  \n",
       "65185                         NaN                NaN  \n",
       "65186                         NaN                NaN  \n",
       "65187                         NaN                NaN  \n",
       "\n",
       "[65188 rows x 15 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.select_dtypes(include=['object'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5e3343",
   "metadata": {},
   "source": [
    "data['CHROM'] must be cleaned as there are ints and strings mixed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6f407138",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',\n",
       "       '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X',\n",
       "       11, 12, 13, 14, 15, 16, 17, 18, 19, 2, 20, 21, 22, 3, 'MT'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['CHROM'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "66ce6cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_map = {2: '2', 3: '3', 11: '11', 12: '12',13: '13',14: '14', 15: '15',16: '16', 17: '17', 18: '18', 19: '19', \n",
    "             20: '20', 21: '21', 22: '22'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "160f4bd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data['CHROM'] = data['CHROM'].replace(chrom_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fd0cfe4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',\n",
       "       '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X',\n",
       "       'MT'], dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['CHROM'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7faca051",
   "metadata": {},
   "outputs": [],
   "source": [
    "for column in data.columns:\n",
    "    if column == 'BLOSUM62':\n",
    "        data[column] = data[column].fillna(method='ffill')\n",
    "    elif data[column].dtype == 'object':\n",
    "        data[column] = data[column].fillna(method='ffill')\n",
    "    elif data[column].dtype == 'float64':\n",
    "        data[column] = data[column].interpolate()\n",
    "\n",
    "data['LoFtool'] = data['LoFtool'].fillna(0)\n",
    "\n",
    "data.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cec55e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.select_dtypes(include=['object']).nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7465b6c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_be_one_hot = ['CLNVC', 'IMPACT', 'SIFT']\n",
    "to_be_mapped = ['SIFT', 'PolyPhen']\n",
    "to_be_encoded = ['REF', 'ALT', 'CLNDN', 'Allele', 'Consequence', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
